core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala

```diff
@@ -37,30 +39,29 @@ class ExpressionConverterSpec extends AnyFlatSpec with OneInstancePerTest with M
   behavior of "Converting arbitrary Spark Expression"
 
   private val dtConverterMock = mock[DataTypeConverter]
-  private val attrConverterMock = mock[AttributeConverter]
-  private val converter = new ExpressionConverter(new DataConverter, dtConverterMock, attrConverterMock)
+  private val converter = new GenericExpressionConverter(new DataConverter, dtConverterMock)
 
   when(dtConverterMock convert NullType -> true) thenReturn nullDataType
   when(dtConverterMock convert StringType -> false) thenReturn stringDataType
 
   it should "support secondary constructor, but only capture params from the primary one" in {
     val expression = new Foo("this parameter should not be captured")
 
-    inside(converter convert expression) {
-      case expr.Generic(name, dataTypeId, children, exprType, Some(params)) =>
+    inside(converter convert(expression, opId)) {
+      case FunctionalExpression(id, Some(dataTypeId), children, extra, name, params) =>
         name shouldEqual "foo"
         dataTypeId shouldEqual nullDataType.id
         children should have size 0
-        exprType shouldEqual classOf[Foo].getName
+        extra should contain("_typeHint" -> "expr.Generic")
         params shouldNot contain key "aParamOfSecondaryConstructor"
     }
   }
 
   it should "ignore properties with null or None value" in {
     val expression = Foo.empty
 
-    inside(converter convert expression) {
-      case expr.Generic(_, _, _, _, Some(params)) =>
+    inside(converter convert(expression, opId)) {
+      case FunctionalExpression(_, _, _, _, _, params) =>
```

Comment:
The number of `_` is growing. Consider replacing an extraction pattern with a type matching pattern followed by an accessor call.